### ISSUE TITLE
error playing live streams/vods

### ISSUE THREAD

**Issue by efirshik** [anxdpanic/plugin.video.twitch#461]
```
2019-11-08 23:42:09.133 T:3571446640 WARNING: script.module.python.twitch: API version |V5| is deprecated, update to |Helix| by |TBD|
2019-11-08 23:42:10.080 T:3571446640  NOTICE: Twitch: Calling |_play| for mode |play| with pos args |[]| and kwargs |{'channel_id': '82875101'}|
2019-11-08 23:42:10.659 T:3571446640   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!                                                     
                                            Error Type: <type 'exceptions.AttributeError'>  
                                            Error Contents: 'dict' object has no attribute 'decode'
                                            Traceback (most recent call last):                     
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 15, in <module>
                                                router.run(sys.argv)                                                                              
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 297, in run
                                                dispatcher.dispatch(mode, queries)                                                                   
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/url_dispatcher.p$
", line 94, in dispatch                                                                                                                                
                                                self.func_registry[mode](*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", lin$
 29, in wrapper                                                                                                                                        
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 157, in _pla$
                                                play.route(twitch_api, seek_time, channel_id, video_id, slug, ask, use_player, quality, channel_name)  
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/routes/play.py", line 116, in
route                                                                                                                                                 
                                                videos = api.get_live(name)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", lin$
 80, in wrapper                                                                                                                                        
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/cache.py", line $
04, in memoizer                                                                                                                                        
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 325, in g$
t_live                                                                                                                                                 
                                                results = self.usher.live(name)
                                              File "/storage/.kodi/addons/script.module.python.twitch/resources/lib/twitch/parser.py", line 51, in m3u$
_wrapper                                                                                                                                               
                                                results = results.decode('utf-8')
                                            AttributeError: 'dict' object has no attribute 'decode'
                                            -->End of Python script error report<--                
2019-11-08 23:42:10.868 T:4092703520   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.twitch/?channel_id=82875101&mo$
e=play]                                              
```

**Comment by neo-neo1** [comment]
I started getting an identical problem just now also. Streams appear, but error out on attempting to play.
```

2019-11-08 14:35:28.275 T:4092644784   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.twitch/?channel_id=23945610&mode=play]
2019-11-08 14:35:30.083 T:3534066560 WARNING: script.module.python.twitch: API version |V5| is deprecated, update to |Helix| by |TBD|
2019-11-08 14:35:30.561 T:3534066560  NOTICE: Twitch: Calling |_play| for mode |play| with pos args |[]| and kwargs |{'channel_id': '136765278'}|
2019-11-08 14:35:31.072 T:3534066560   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.AttributeError'>
                                            Error Contents: 'dict' object has no attribute 'decode'
                                            Traceback (most recent call last):
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 15, in <module>
                                                router.run(sys.argv)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 297, in run
                                                dispatcher.dispatch(mode, queries)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/url_dispatcher.py", line 94, in dispatch
                                                self.func_registry[mode](*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", line 29, in wrapper
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 157, in _play
                                                play.route(twitch_api, seek_time, channel_id, video_id, slug, ask, use_player, quality, channel_name)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/routes/play.py", line 116, in route
                                                videos = api.get_live(name)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", line 80, in wrapper
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/cache.py", line 104, in memoizer
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 325, in get_live
                                                results = self.usher.live(name)
                                              File "/storage/.kodi/addons/script.module.python.twitch/resources/lib/twitch/parser.py", line 51, in m3u8_wrapper
                                                results = results.decode('utf-8')
```

**Comment by aGreatDay** [comment]
I'm having the same issue
 

> 2019-11-08 22:04:16.854 T:1480585952   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.AttributeError'>
                                            Error Contents: 'dict' object has no attribute 'decode'
                                            Traceback (most recent call last):
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 15, in <module>
                                                router.run(sys.argv)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 297, in run
                                                dispatcher.dispatch(mode, queries)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/url_dispatcher.py", line 94, in dispatch
                                                self.func_registry[mode](*args, **kwargs)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", line 29, in wrapper
                                                result = func(*args, **kwargs)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 157, in _play
                                                play.route(twitch_api, seek_time, channel_id, video_id, slug, ask, use_player, quality, channel_name)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/routes/play.py", line 116, in route
                                                videos = api.get_live(name)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", line 80, in wrapper
                                                result = func(*args, **kwargs)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/cache.py", line 104, in memoizer
                                                result = func(*args, **kwargs)
                                              File "/home/pi/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 325, in get_live
                                                results = self.usher.live(name)
                                              File "/home/pi/.kodi/addons/script.module.python.twitch/resources/lib/twitch/parser.py", line 51, in m3u8_wrapper
                                                results = results.decode('utf-8')
                                            AttributeError: 'dict' object has no attribute 'decode'
                                            -->End of Python script error report<--
2019-11-08 22:04:17.116 T:1902371680   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.twitch/?channel_id=15564828&mode=play]

**Comment by Warloxx** [comment]
Can confirm, observing same issue as above. Although I was trying to watch a VOD rather than livestream.
```
2019-11-08 21:36:07.030 T:1344902000   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.AttributeError'>
                                            Error Contents: 'dict' object has no attribute 'decode'
                                            Traceback (most recent call last):
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/addon_runner.py", line 15, in <module>
                                                router.run(sys.argv)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 311, in run
                                                dispatcher.dispatch(mode, queries)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/url_dispatcher.py", line 94, in dispatch
                                                self.func_registry[mode](*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", line 29, in wrapper
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/router.py", line 171, in _play
                                                play.route(twitch_api, seek_time, channel_id, video_id, slug, ask, use_player, quality, channel_name)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/routes/play.py", line 87, in route
                                                _videos = api.get_vod(video_id)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/error_handling.py", line 80, in wrapper
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/common/cache.py", line 104, in memoizer
                                                result = func(*args, **kwargs)
                                              File "/storage/.kodi/addons/plugin.video.twitch/resources/lib/twitch_addon/addon/api.py", line 326, in get_vod
                                                results = self.usher.video(video_id)
                                              File "/storage/.kodi/addons/script.module.python.twitch/resources/lib/twitch/parser.py", line 51, in m3u8_wrapper
                                                results = results.decode('utf-8')
                                            AttributeError: 'dict' object has no attribute 'decode'
                                            -->End of Python script error report<--
```

**Comment by aGreatDay** [comment]
Streamlink is having the same issue: https://github.com/streamlink/streamlink/issues/2680

**Comment by sth** [comment]
Enabling debug logging gives some insight:

```
2019-11-08 22:32:04.256 T:140712499615488   DEBUG: Twitch: Calling cached method: |twitch_addon.addon.api.Twitch.get_
live|
2019-11-08 22:32:04.258 T:140712499615488   DEBUG: script.module.python.twitch: GET QUERY: url: |https://api.twitch.tv/api/channels/heromarine/access_token|, params: |{'need_https': 'true', 'platform': 'web', 'player_backend': 'mediaplayer'}|, data: |{}|, headers: |{'Client-ID': '*****', 'Authorization': 'OAuth ***********'}|, target_func: |channel_token|
2019-11-08 22:32:04.259 T:140712499615488   DEBUG: script.module.python.twitch: Downloading: |https://api.twitch.tv/api/channels/heromarine/access_token?need_https=true&platform=web&player_backend=mediaplayer|
2019-11-08 22:32:04.566 T:140712499615488   DEBUG: script.module.python.twitch: url: |https://api.twitch.tv/api/channels/heromarine/access_token| parameters: |{'need_https': 'true', 'platform': 'web', 'player_backend': 'mediaplayer'}|
                                            {
                                                "error": "Gone", 
                                                "message": "this API has been removed.", 
                                                "status": 410
                                            }
2019-11-08 22:32:04.571 T:140712499615488   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
```

So it seems like the used API call is not available like this anymore.

**Comment by Zydox** [comment]
From what I saw in the streamlink forum, it seems like Twitch is switching to their new API called helix:
https://dev.twitch.tv/docs/api

**Comment by neo-neo1** [comment]
Anyone know of a workaround?

**Comment by fragande** [comment]
> 
> 
> Anyone know of a workaround?

I managed to get it working by changing the `OAuth client ID` under the Developer menu to the Twitch website Client ID (`kimne78kx3ncx6brgo4mv6wki5h1ko`) and using `OAuth Token` extracted from web browser cookie (`auth-token`). This sort of works as streams are playing again but I don't get the user categories (e.g. Followed channels and such).

Lets hope this isn't Twitch shutting down 3rd party applications :(

**Comment by prometheanfire** [comment]
I had to remove my existing auth-token, then change the client ID to the twitch website one, then pull the new auth-token from the website manually (not use the built in request-auth-token function).  If I did not remove my existing auth token the change in the client ID would be rejected (probably because the token was associated with the v3 api?)

To me it looks like client IDs and tokens associated with the v3 api are what are not working.

**Comment by ghost** [comment]
Neither of the above worked for me, but I might have used the wrong token. That being said even if I have no auth info at all trying to play a stream is erroring for me.

**Comment by Wibbbs** [comment]
Seeing the same error here. Hopefully this isn't the end of the v3 api.

**Comment by Wibbbs** [comment]
Its working again.

**Comment by xenius** [comment]
Working again here. With the client-id empty. Just using the 0authtoken. Generated by twitchaddon.page/***** link. 

EDIT: And broken again?

**Comment by kuraikaze** [comment]
I was getting the same API fail and/or timeout issues when Twitch tied itself in knots dealing with the huge extra load from the Fortnite new season event a few weeks ago.
Is there a big load-generating event happening on the service currently ?

**Comment by efirshik** [comment]
"following live channels" work again, without manipulation with token :)

**Comment by fragande** [comment]
Problem back today again. Seems that there's not much to be done other than hope Twitch doesn't break the undocumented API endpoint permanently. From the discussion over at streamlink it seems there's no real alternative.

Question is would it be possible to use the Twitch client ID with your web user auth-token and still get Followed channels and similar in Twitch-on-Kodi?

**Comment by efirshik** [comment]
yes problem back again today :(

**Comment by neo-neo1** [comment]
> Problem back today again. Seems that there's not much to be done other than hope Twitch doesn't break the undocumented API endpoint permanently. From the discussion over at streamlink it seems there's no real alternative.
> 
> Question is would it be possible to use the Twitch client ID with your web user auth-token and still get Followed channels and similar in Twitch-on-Kodi?

Confirmed, same problem is back. I hope this is fixable...

**Comment by prometheanfire** [comment]
I just streamed something using the 'fix' mentioned previously (their client-id, my auth token).

**Comment by neo-neo1** [comment]
> I just streamed something using the 'fix' mentioned previously (their client-id, my auth token).

Indeed, I can confirm this is works though is not the ideal solution. Credit goes to @fragande for detailed instructions

**Comment by fragande** [comment]
Yeah, it's a workaround for now I guess. Today when I changed back to the Twitch client ID + web auth token I actually didn't lose Followed channels, no idea why. According to the cookie data the web auth token should be good for 6 months. 

If it's possible to keep Followed channels and such it could be a "permanent" solution (until Twitch shuts that down too ...), albeit a very dirty one that probably breaks ToS. Third party apps are supposed to use their own client ID's.

> 
> 
> I just streamed something using the 'fix' mentioned previously (their client-id, my auth token).

> 
> 
> Indeed, I can confirm this is works though is not the ideal solution. Credit goes to @fragande for detailed instructions

Did either of you get to keep Followed channels when doing this?

**Comment by neo-neo1** [comment]
> Yeah, it's a workaround for now I guess. Today when I changed back to the Twitch client ID + web auth token I actually didn't lose Followed channels, no idea why. According to the cookie data the web auth token should be good for 6 months.
> 
> If it's possible to keep Followed channels and such it could be a "permanent" solution (until Twitch shuts that down too ...), albeit a very dirty one that probably breaks ToS. Third party apps are supposed to use their own client ID's.
> 
> > I just streamed something using the 'fix' mentioned previously (their client-id, my auth token).
> 
> > Indeed, I can confirm this is works though is not the ideal solution. Credit goes to @fragande for detailed instructions
> 
> Did either of you get to keep Followed channels when doing this?

Surprisingly enough yes, the followed channels do work as expected

**Comment by Baerht** [comment]
this issue seems to be wide spread across several 3 party viewing 'apps' . Streamview Is also experiencing this as well

**Comment by fragande** [comment]
> 
> 
> Surprisingly enough yes, the followed channels do work as expected

I guess I did something differently the first time then. In theory I don't think there should be a difference as it's the same data Twitch-on-Kodi is authorized to access when creating the OAuth, but didn't work for me first time around.

> 
> 
> this issue seems to be wide spread across several 3 party viewing 'apps' . Streamview Is also experiencing this as well

Yes, it's affecting every 3rd party app using this API endpoint. Seems Twitch are periodically blocking every client ID but their own for unknown reasons.  It seems Twitch aren't happy about this endpoint being used in the first place as it's undocumented and unsupported but they haven't cracked down on the usage of it until now AFAIK.

**Comment by ghost** [comment]
I noticed a few versions of the Twitch API in the developer section of their site. I don't know much about this, or other addons, but how difficult would it be to use a newer more supported API version? I get there would be work involved but would it be worth it for hopefully some future proofing?

**Comment by xenius** [comment]
Can anyone explain the fix workaround? When i use twitch id oauthtoken (kimne78kx3ncx6brgo4mv6wki5h1ko)and the web generated auth token ,its still not working?

**Comment by anxdpanic** [comment]
As mentioned, this is due to an undocumented API endpoint and is not included on their developer site.

We are currently use v5 at the moment for all other functions, which currently has no eol date. The helix API is ready in the module for when the time comes, it comes with additional limitations which is why we haven't migrated to helix yet.

Currently the only way to temporarily workaround this is to use the client id and oauth token from the website.

**Comment by anxdpanic** [comment]
@xenius 
Just double checked here and it is working here. I'd double check you didn't get an extra character or something in the oauth token, and it shouldn't be required but you could also run Settings - Maintenance - Reset function cache

**Comment by xenius** [comment]
The cliend id (kimne78kx3ncx6brgo4mv6wki5h1ko) and my generated auth token is not working....

**Comment by anxdpanic** [comment]
you want to use the oauth token you see in the same place as you found the client id, not the one you generated from the add-on (back that one up first).

**Comment by xenius** [comment]
Where can i find the exact 2 things i need to get this working again?

**Comment by neo-neo1** [comment]
> Where can i find the exact 2 things i need to get this working again?

Read this post, instructions have been posted.

**Comment by xenius** [comment]
> > Anyone know of a workaround?
> 
> I managed to get it working by changing the `OAuth client ID` under the Developer menu to the Twitch website Client ID (`kimne78kx3ncx6brgo4mv6wki5h1ko`) and using `OAuth Token` extracted from web browser cookie (`auth-token`). This sort of works as streams are playing again but I don't get the user categories (e.g. Followed channels and such).
> 
> Lets hope this isn't Twitch shutting down 3rd party applications :(

If this is the post u mean. I still dont get it though
... How do i extract that from my browser cookie?

**Comment by fragande** [comment]
> 
> 
> Where can i find the exact 2 things i need to get this working again?

You set the Twitch client ID (kimne78kx3ncx6brgo4mv6wki5h1ko) under `Developer > OAuth Client ID` in the GUI. For the `Login > OAuth Token` you want to extract it from the web page cookie (it's **not** the OAuth generated by the add-on you want). In Firefox go to twitch.tv (be logged in), right click anywhere and click `Inspect element`, go to the `Storage` tab. There should be a post called `auth-token`. Copy that value and enter it in Kodi `Login > OAuth Token`.

To avoid having to type everything manually you can also edit the `settings.xml` file in `.kodi/userdata/addon_data/plugin.video.twitch`:

```
<setting id="oauth_clientid">kimne78kx3ncx6brgo4mv6wki5h1ko</setting>
<setting id="oauth_token">paste_your_web_auth-token_here</setting>
```

**Comment by xenius** [comment]
Thank you... Working for me now. And i still have ly followed channels.

**Comment by usphil** [comment]
I can not find the Oauth_ClientID, what the browser you use? Thanks!

**Comment by fragande** [comment]
> 
> 
> I can not find the Oauth_ClientID, what the browser you use? Thanks!

The client ID you want to use is `kimne78kx3ncx6brgo4mv6wki5h1ko`, it's Twitch's own client ID. This value is static and same for everyone. Just use it in combination with your personal `auth-token` from the web cookie (see my post above for Firefox instructions, for other browsers I don't know but it shouldn't be hard to find).

**Comment by usphil** [comment]
Thank you so much. It's working now

**Comment by pitsi** [comment]
Thank you for all the info and the workaround. I do not have a twitch account, so I can not get an`auth-token` for myself. Can I use only the `clientid` provided above or will I run into problems?
I just entered it on my libreelec installation and I will check tonight if it works.

**Comment by ghost** [comment]
### How to solve this error?

2019-11-12 13:06:54.736 T:16140 WARNING: script.module.python.twitch: API version |V5| is deprecated, update to |Helix| by |TBD|
2019-11-12 13:06:54.955 T:16140  NOTICE: Twitch: Calling |_list_followed| for mode |followed| with pos args |['channels']| and kwargs |{}|
2019-11-12 13:06:57.486 T:16140  NOTICE: Twitch: Calling |_list_channel_video_categories| for mode |channel_videos| with pos args |[]| and kwargs |{'channel_id': '23798553', 'display_name': 'MJRAMON', 'channel_name': 'mjramon'}|
2019-11-12 13:06:58.243 T:16140  NOTICE: Twitch: Calling |_list_channel_videos| for mode |channel_video_list| with pos args |['archive']| and kwargs |{'channel_id': '23798553', 'game': None}|
2019-11-12 13:07:00.409 T:16140  NOTICE: Twitch: Calling |_play| for mode |play| with pos args |[]| and kwargs |{'video_id': 'v506992318'}|
2019-11-12 13:07:02.750 T:16140   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.AttributeError'>
                                            Error Contents: 'dict' object has no attribute 'decode'
                                            Traceback (most recent call last):
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\addon_runner.py", line 15, in <module>
                                                router.run(sys.argv)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\router.py", line 297, in run
                                                dispatcher.dispatch(mode, queries)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\addon\common\url_dispatcher.py", line 94, in dispatch
                                                self.func_registry[mode](*args, **kwargs)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\addon\error_handling.py", line 29, in wrapper
                                                result = func(*args, **kwargs)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\router.py", line 157, in _play
                                                play.route(twitch_api, seek_time, channel_id, video_id, slug, ask, use_player, quality, channel_name)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\routes\play.py", line 87, in route
                                                _videos = api.get_vod(video_id)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\addon\error_handling.py", line 80, in wrapper
                                                result = func(*args, **kwargs)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\addon\common\cache.py", line 104, in memoizer
                                                result = func(*args, **kwargs)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\plugin.video.twitch\resources\lib\twitch_addon\addon\api.py", line 314, in get_vod
                                                results = self.usher.video(video_id)
                                              File "C:\Users\Prototype\AppData\Roaming\Kodi\addons\script.module.python.twitch\resources\lib\twitch\parser.py", line 51, in m3u8_wrapper
                                                results = results.decode('utf-8')
                                            AttributeError: 'dict' object has no attribute 'decode'
                                            -->End of Python script error report<--
2019-11-12 13:07:02.797 T:9072   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.twitch/?video_id=v506992318&mode=play]
2019-11-12 13:07:13.547 T:9072  NOTICE: Stopping player

**Comment by fragande** [comment]
> 
> ### How to solve this error?
> 

Read this thread, the only workaround as for now is to replace client ID and use web auth token. See https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552670340.

**Comment by Baerht** [comment]
looks like they have unblocked things ,for now , 11:00 PST

**Comment by Jhyrachy** [comment]
Locked again

**Comment by RevAngel7** [comment]
This is a quote. Original article is found at https://discuss.dev.twitch.tv/t/twitch-api-v3-and-v5-deprecation-updates/19144

>  

> Twitch API v3 and v5 Deprecation Updates
> Announcements
> jbulava
> Twitch Staff
> 1
> Dec '18
> 
> Update (2019/07/25): Please see our latest information regarding the shutdown timeline for v3 here 241.
> 
> Greetings from Twitch,
> 
> We have important updates to share regarding deprecation of v3 and v5 of the Twitch API.
> 
> As previously communicated 54 in February 2017 and documented in the v5 API guide 107, Twitch planned to remove third-party support for the v3 API on December 31, 2018. However, we are delaying this process by a few weeks to provide developers more time to prepare for the upcoming changes. Starting in late January 2019, Twitch will begin a controlled shutdown of v3 over several weeks and withdraw v3 from public third-party access in early 2019. More details on the shutdown process and exact timelines will be made available in mid-January 2019. We will communicate these changes prior to starting shutdown and withdrawal activities through our developer channels like the Developer Forums, so please stay tuned.
> 
> Also, we previously documented that API calls without a specific API version in the headers would be directed to v5 endpoints after December 31, 2018. This change is also on hold. We will confirm an exact date for this transition in mid-January.
> 
> Finally, if your existing Twitch integration uses v3 of the Twitch API, you should immediately migrate to the new Twitch API. Documentation for the new Twitch API is here 132. If you have questions about moving to the new Twitch API, visit the API section 63 on our Twitch Developer Forums to learn more and ask questions.
> 
> As an important reminder, v5 of the Twitch API remains on a deprecation path. Twitch will announce an end-of-support date and shutdown plan for v5 in 2019.
> 
> We are working to make sure the upcoming changes will be as well communicated and documented as possible. As members of the Twitch community, we understand that you count on the reliability and stability of Twitch APIs. It is our goal, through these changes, to meet these requirements so that you can build more amazing experiences on Twitch. Thank you for your continued support and patience.

**Comment by RevAngel7** [comment]
This is also a quote. Source is found at https://discuss.dev.twitch.tv/t/twitch-api-v3-shutdown-timeline/21931

>  Twitch API v3 Shutdown Timeline
> Announcements
> jbulava
> Twitch Staff
> Jul 25
> 
> On November 15, 2016, Twitch announced that the v3 API had been deprecated 43 with the release of v5. Today, we’re moving forward with a controlled shutdown 43 of v3 and would like to share our plans.
> 
> Starting on August 12, 2019, we will begin implementing scheduled outage windows of v3 endpoints over several weeks ending with a complete shutdown of v3 endpoints on September 12, 2019. If you do not take action to migrate your application to v5 of the API, your application will experience outages during this time and will break on September 12. The expected scheduled outage windows are as follows:
> 
>     August 12 – 15 minute shutdown starting at 11:00am Pacific 9
>     August 15 – 1 hour shutdown starting at 11:00am Pacific 4
>     August 21 – 6 hour shutdown starting at 11:00am Pacific 13
>     August 22 – 24 hour shutdown starting at 11:00am Pacific 29
>     September 5 – 24 hour shutdown starting at 11:00am Pacific 16
>     September 12 – Permanent shutdown starting at 11:00am Pacific 16
> 
> We have prepared a migration guide 242 to assist you with updates to v5. We have also created a dedicated category on the Twitch Developer forums for migration questions 46. And as always, you can talk directly with other members of the community on the TwitchDev Discord server 18.
> 
> Thank you for being a valued member of the Twitch Developer community. This change is part of our continued effort to improve developer products which enable creators like yourself to deliver the interactive experiences that make Twitch what it is today.

**Comment by anxdpanic** [comment]
@RevAngel7 
From this issue 
https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552665286
https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552670340

**Comment by RevAngel7** [comment]
> @RevAngel7
> From this issue
> [#461 (comment)](https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552665286)
> [#461 (comment)](https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552670340)

Hello anxdpanic,
I got the error message "wrong oauth token" twice after I changed the oauth token, but it worked then. 
Thank you for your help!

**Comment by pokemane** [comment]
[This specific issue has nothing to do with the v3/v5 API.](https://github.com/streamlink/streamlink/issues/2680#issuecomment-552594488) It uses a "private" API endpoint that isn't documented, and Twitch haven't communicated anything about it.

The Streamlink thread, and the progress/updates that team are making, should be pretty much in line with what's required to make the kodi plugin work.

**Comment by apo86** [comment]
Maybe a stupid question, but what's the oauth token even do on this specific API call?

I set the client ID to Twitch's and then straight up removed the [oauth authorization header from the call in the twitch api module](https://github.com/MrSprigster/script.module.python.twitch/blob/master/resources/lib/twitch/queries.py#L158#L159).

Everything as far as I can tell works fine. Still have the addon configured with the same oauth token I generated ages ago. My followed channels are still there, I can follow/unfollow, live notifications work, chat works, vods work...

**Comment by efirshik** [comment]
today https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552670340 trick not work.

`2019-11-17 13:06:01.541 T:3598697328   ERROR: ResourceUnavailableException: script.module.python.twitch: GET Query to https://api.twitch.tv/kraken/, params {}, data {},  headers {'Client-ID': u'kimne78kx3ncx6brgo4mv6wki5h1ko', 'Authorization': 'OAuth ***********', 'ACCEPT': 'application/vnd.twitchtv.v5+json', 'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:6.0) Gecko/20100101 Firefox/6.0'}`

**Comment by QueenGit** [comment]
I managed it to get it working again, by changing the OAuth client ID under the Developer menu to the Twitch website Client ID (kimne78kx3ncx6brgo4mv6wki5h1ko). 
I DONT enter any OAuth Token.

**Comment by the-isz** [comment]
Confirming @QueenGit: Leaving the oauth_token setting empty still gets me a working Twitch addon.

What I had to learn the hard way: kodi (or the Twitch addon?) doesn't like XML comments and crashes :(

**Comment by mrscotland** [comment]
https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552670340
works for me. thanks!

**Comment by MrSprigster** [comment]
@anxdpanic it seems omitting the oauth token in the live_request and _live headers in usher.py in python twitch seems to fix this issue.

**Comment by anxdpanic** [comment]
Without the oauth token in those areas, there is no subscriber only streams, subscriber qualities or subscriber vods. Will check if there is anything to make it smoother in the mean time.

**Comment by MrSprigster** [comment]
True, maybe a toggle in Developer settings,  Disable Playback Authentication (Workaround)

**Comment by anxdpanic** [comment]
Actually working on something atm.
I really don't want to hard code their client id as they will likely just change it at that point, so I added a 'Private API Credentials' to Settings -> Login

There you can put in the web pages client id and oauth token(for users who want subscriber only content access), they will only be used on those end points. 

So regular auth credentials can be left in place, and if Twitch happens to revert the change users can just delete the private credentials.

**Comment by anxdpanic** [comment]
The new zips are ready, https://github.com/MrSprigster/Twitch-on-Kodi/releases/tag/2.4.7
You'll need to install first ```script.module.python.twitch-2.0.10.zip``` before installing ```plugin.video.twitch-2.4.7.zip```

If you've already changed your client id and/or oauth token, you can safely remove the client id and use your old oauth token. If you have the oauth token from the web page, you can add that to ```Settings - Login - Private API Credentials - OAuth Token```.

If the client id changes from ```kimne78kx3ncx6brgo4mv6wki5h1ko``` then you can update ```Settings - Login - Private API Credentials - OAuth Client-ID``` This also means the oauth token likely changed as well, so update that as well if necessary.

You can also edit the `settings.xml` file
```
    <setting id="private_oauth_clientid">paste_client_id_from_the_web_here</setting>
    <setting id="private_oauth_token">paste_oauth_token_from_the_web_here</setting>
```

**Comment by anxdpanic** [comment]
I updated the previous comment with new zips, 2.4.7 and 2.0.10. Following Games, Toggle Follow on Games started responding with the `Gone` response shortly after the last release. Those should now work with the private credentials entered as well.

**Comment by EiswolfHS** [comment]
Yeah it works again! Thanks!

**Comment by Extarys** [comment]
I updated one of my Kodis yesterday. Thanks for the hard and fast work. I wonder if I'm the only one having issues accessing the "Game" menu? I can see the featured streams but not the Games section.

I cannot provides logs at this time. I'll investigate this evening or tomorrow.

**Comment by efirshik** [comment]
please fix "get oauth token" link in new version
![Screenshot from 2019-11-23 21-06-33](https://user-images.githubusercontent.com/43796942/69482467-3694d300-0e35-11ea-832a-4fe141163062.png)

**Comment by anxdpanic** [comment]
@efirshik it is case sensitive https://twitchaddon.page.link/eetJ

**Comment by Dr0plexx** [comment]
I think i need some help. On my Smart-TV with Kodi installed it doesnt work at all since Twitch changed the API or whatever. When i enter my OAtuth Token, he shows me the correct Channels i follow, but when i click a Stream to watch there is a popup with "Gone [410] This API has been removed." And later he say me: Invalid OAutch Token. 
Its on TV, i cant read log or edit a settings file. :(

**Comment by Extarys** [comment]
@Dr0plexx you can read @anxdpanic comments above.

> The new zips are ready, https://github.com/MrSprigster/Twitch-on-Kodi/releases/tag/2.4.7
> You'll need to install first script.module.python.twitch-2.0.10.zip before installing plugin.video.twitch-2.4.7.zip

**Comment by Dr0plexx** [comment]
Thanks, but i installed them already. Those Errors are "made" with the Pyton 2.0.10 and Kodi 2.4.7.
I actually need to use the sh*ty Twitch App on TV for watching Streams. :/

**Comment by Extarys** [comment]
Oh, sorry :/
I don't really know what to say. For me using LibreElec, the Featured streams works fine, but everything else gives me an error, which I can cope with for the time being.
Maybe someone else will stop by and have some insight for you though. Hang tight.

**Comment by anxdpanic** [comment]
Would need complete logs with debug logging enabled to see what is going on. It is and has been working with all features for me.

**Comment by Dr0plexx** [comment]
Thats the problem. I dont know how and where to find those loggings on TV. Maybe its impossible to reach those files because of very limited access and the lack of a File Browser or something like that.
Is it normal the OAuth Token change every time i click "Get OAuth Token"? I thought its unique.

**Comment by anxdpanic** [comment]
Yes it would be different every time, you're generating a new one every time that is done.

https://kodi.wiki/view/Log_file/Easy <- Follow this, and use step 3.1 for uploading the log

**Comment by Dr0plexx** [comment]
Thats nice. Here it is:
https://paste.kodi.tv/oxazeyasud

**Comment by anxdpanic** [comment]
That's just a regular log, need to do Step 2 first

**Comment by Dr0plexx** [comment]
Sorry. 
I enabled debug logging and restartet Kodi. This is the new one. Hope thats the right one.
https://paste.kodi.tv/xemogetewo

**Comment by anxdpanic** [comment]
That was the one. Looks like it is using the wrong Client ID

Can you double check that `Settings -> Login -> Private API Credentials -> OAuth Client ID` is `kimne78kx3ncx6brgo4mv6wki5h1ko`
and `Settings -> Login -> Private API Credentials -> OAuth Token` is blank. Unless you get the oauth token from inspecting the web page, as mentioned https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-552670340 the information gathered here goes in a new setting as mentioned https://github.com/MrSprigster/Twitch-on-Kodi/issues/461#issuecomment-555186911

Also `Settings -> Developer` should be blank.

**Comment by Dr0plexx** [comment]
Wow that kimne.. Client ID thing was the only Problem? It works now with my OAuth Token i created in the first row and the kimne.. ID in the Private API. I have tried that some days ago. But maybe in the wrong row. I am sorry for my dumbness :)
But really THANKS for this great help and support!

**Comment by Tony322** [comment]
Hmm. Updated Python script, updated the addon. Now I dont get an error, but the stream starts loading for a long time then i get "One or more items failed to play". Tried unisntalling and reinstalling from scratch, but got the same issue. Clientid is correct. Also tried with and without my oauth token. Is it only me?

**Comment by anxdpanic** [comment]
2.4.7 was merged into the official repository, will be available when caching allows

**Comment by BlackEdder** [comment]
> You set the Twitch client ID (kimne78kx3ncx6brgo4mv6wki5h1ko) under `Developer > OAuth Client ID` in the GUI. For the `Login > OAuth Token` you want to extract it from the web page cookie (it's **not** the OAuth generated by the add-on you want). In Firefox go to twitch.tv (be logged in), right click anywhere and click `Inspect element`, go to the `Storage` tab. There should be a post called `auth-token`. Copy that value and enter it in Kodi `Login > OAuth Token`.
> 
> To avoid having to type everything manually you can also edit the `settings.xml` file in `.kodi/userdata/addon_data/plugin.video.twitch`:
> 
> ```
> <setting id="oauth_clientid">kimne78kx3ncx6brgo4mv6wki5h1ko</setting>
> <setting id="oauth_token">paste_your_web_auth-token_here</setting>
> ```

For others who might run into the same issue as me. I ended up having to change the clientid as well, because streams I subscribe to (so I can watch their past broadcasts) stopped working for me. To get the new clientid, I used firefox as above and copy and pasted the contents of the `api_token` into the clientid field (as well as copying the "auth-token" into the token field).
